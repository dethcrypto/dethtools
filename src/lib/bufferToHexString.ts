export function bufferToHexString(byteArray: Buffer): string {
  return Array.from(byteArray, function (byte) {
    return ('0' + (byte & 0xff).toString(16)).slice(-2);
  }).join('');
}
