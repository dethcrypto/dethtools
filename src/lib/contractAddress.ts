import { getContractAddress as getContractAddressImpl } from '@ethersproject/address';

export function getContractAddress(params: {
  from: string;
  nonce: string;
}): string {
  return getContractAddressImpl(params);
}
