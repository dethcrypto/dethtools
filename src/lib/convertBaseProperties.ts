export const base = ['binary', 'octal', 'decimal', 'hexadecimal'] as const;

export type Base = typeof base[number];
