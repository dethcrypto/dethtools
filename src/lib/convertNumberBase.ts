import BigNumber from 'bignumber.js';

import { NumberBaseType } from './numberBaseConvertProperties';

export function convertNumberBase(
  value: string,
  fromUnit: NumberBaseType,
  toUnit: NumberBaseType,
): string | undefined {
  for (const [fromBaseName, fromBaseNumber] of Object.entries(
    numberBaseConversionDict,
  )) {
    if (fromBaseName === fromUnit) {
      for (const [toBaseName, toBaseNumber] of Object.entries(
        numberBaseConversionDict,
      )) {
        if (toBaseName === toUnit) {
          return convertNumberBaseWithBn(
            value,
            toBaseName,
            fromBaseNumber,
            toBaseNumber,
          );
        }
      }
    }
  }
}

// @internal
const numberBaseConversionDict = {
  binary: 2,
  octal: 8,
  decimal: 10,
  hexadecimal: 16,
};

// @internal
function convertNumberBaseWithBn(
  value: string,
  toBaseName: string,
  from: number,
  to: number,
): string {
  if (toBaseName === 'octal') {
    return '0' + new BigNumber(value, from).toString(to);
  } else if (toBaseName === 'hexadecimal') {
    return '0x' + new BigNumber(value, from).toString(to);
  } else {
    return new BigNumber(value, from).toString(to);
  }
}
