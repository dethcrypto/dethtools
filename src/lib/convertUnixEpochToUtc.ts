import assert from 'assert';

import { unsafeEntries } from '../../src/misc/unsafeEntries';

export function convertUnixEpochToUtc(unixEpoch: string):
  | {
      utcDate: string;
      unixTimestampFormat: UnixTimestampFormat;
    }
  | undefined {
  const unixTimestampFormat = detectUnixTimestampFormat(unixEpoch);
  assert(
    unixTimestampFormat,
    `The value ${unixEpoch} doesn't fit into any known format range`,
  );
  const utcDate = convertUnixEpochToUtcByDate(unixTimestampFormat, unixEpoch);
  if (utcDate) {
    return {
      utcDate,
      unixTimestampFormat,
    };
  }
}

export type UnixTimestampFormat =
  | 'seconds'
  | 'milliseconds'
  | 'microseconds'
  | 'nanoseconds';

// @internal
function convertUnixEpochToUtcByDate(
  unixTimestampFormat: UnixTimestampFormat,
  unixEpoch: string,
): string | undefined {
  return new Date(
    Number(unixEpoch) *
      unixTimestampFormatToEpochMultiplier[unixTimestampFormat],
  ).toUTCString();
}

// @internal
function detectUnixTimestampFormat(
  unixEpoch: string,
): UnixTimestampFormat | undefined {
  const len = unixEpoch.length;
  for (const [unixTimestampFormat, range] of unsafeEntries(
    unixTimestampFormatToLengthRange,
  )) {
    if (len <= range[1] && len >= range[0]) {
      return unixTimestampFormat;
    }
  }
}

// @internal
const unixTimestampFormatToLengthRange: Record<
  UnixTimestampFormat,
  [number, number]
> = {
  seconds: [1, 11],
  milliseconds: [12, 14],
  microseconds: [15, 16],
  nanoseconds: [17, 21],
};

// @internal
const unixTimestampFormatToEpochMultiplier = {
  milliseconds: 1,
  seconds: 1_000,
  microseconds: 1 / 1_000,
  nanoseconds: 1 / 1_000_000,
};
