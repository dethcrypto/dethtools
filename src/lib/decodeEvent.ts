import {
  EventFragment,
  FormatTypes,
  Interface,
  ParamType,
} from '@ethersproject/abi';
import { version } from '@ethersproject/abi/lib/_version';
import { BigNumber } from '@ethersproject/bignumber';
import { Logger } from '@ethersproject/logger';

import { combinations as generateCombinations } from '../misc/combinations';

const logger = new Logger(version);

export interface EventProps {
  topics: string[];
  data: string;
}

interface AbstractDecodedEventResult {
  eventFragment: EventFragment;
  hasArgumentNames: boolean;
  signature: string;
  fullSignature: string;
}

export interface DecodedEventResultWithArgNames
  extends AbstractDecodedEventResult {
  hasArgumentNames: true;
  args: Record<string, string | BigNumber>;
}

export interface DecodedEventResultWithoutArgNames
  extends AbstractDecodedEventResult {
  hasArgumentNames: false;
  args: ReadonlyArray<string | BigNumber>;
}

export type DecodedEventResult =
  | DecodedEventResultWithArgNames
  | DecodedEventResultWithoutArgNames;

export function decodeEvent(
  iface: Interface,
  eventProps: EventProps,
): DecodedEventResult {
  try {
    const decoded = iface.parseLog(eventProps);
    /**
     * Ethers under the hood converts EventFragment to object-like form without format function, so we rehydrate it here.
     */
    const eventFragment = EventFragment.fromObject({
      ...decoded.eventFragment,
      _isFragment: false, // Otherwise this function noops
    });
    const hasArgumentNames = doesContainNamedKeys(decoded.args);
    const decodedEventResult = {
      args: hasArgumentNames ? omitNonNamedKeys(decoded.args) : decoded.args,
      eventFragment: eventFragment,
      hasArgumentNames,
      signature: decoded.signature,
      fullSignature: eventFragment.format(FormatTypes.full),
    };
    return decodedEventResult as DecodedEventResult;
  } catch (e: any) {
    switch (e.reason) {
      case 'no matching event':
        throw e;
      case 'data out-of-bounds':
        // At this point we know that we have a matching topic, but the number of indexed args is incorrect
        return tryToDecodeEvent(iface, eventProps);
      default:
        throw e;
    }
  }
}

/**
 * We assume that we don't have partial indexing information.
 * This may not be true, but it's an edge case and it's PITA to handle.
 */
export function tryToDecodeEvent(
  iface: Interface,
  eventProps: EventProps,
): DecodedEventResult {
  const indexedArgsCount = eventProps.topics.length - 1;
  const topichash = eventProps.topics[0];

  for (const name in iface.events) {
    if (topichash === iface.getEventTopic(name)) {
      const event = iface.getEvent(name);
      const inputsCombinations = combinationsMap(
        event.inputs,
        indexedArgsCount,
        (input: ParamType): ChangedParamType => ({
          arrayChildren: input.arrayChildren,
          components: input.components,
          name: input.name,
          type: input.type,
          arrayLength: input.arrayLength,
          baseType: input.baseType,
          indexed: true,
        }),
      );

      let error: unknown;
      for (const inputs of inputsCombinations) {
        const newEvent = EventFragment.fromObject({
          inputs,
          anonymous: event.anonymous,
          type: event.type,
          name: event.name,
        });

        iface.events[name] = newEvent;

        try {
          const args = iface.decodeEventLog(
            newEvent,
            eventProps.data,
            eventProps.topics,
          );

          return {
            args,
            hasArgumentNames: false,
            eventFragment: newEvent,
            signature: newEvent.format(FormatTypes.sighash),
            fullSignature: newEvent.format(FormatTypes.full),
          };
        } catch (err) {
          error = err;
          // We expect this to fail for some combinations. It's only a problem if it fails for all combinations.
        }
      }

      if (error) {
        throw error;
      }
    }
  }

  return logger.throwArgumentError('no matching event', 'topichash', topichash);
}

// @internal
export function omitNonNamedKeys<T extends object>(
  object: T,
): Record<string, string | BigNumber> | T {
  const obj: Record<string, string | BigNumber> = {};
  for (const [key, val] of Object.entries(object)) {
    if (isNaN(parseInt(key))) {
      obj[key] = val;
    }
  }
  return obj;
}

// @internal
export function doesContainNamedKeys<T extends object>(object: T): boolean {
  return Object.keys(object).some((key) => isNaN(parseInt(key)));
}

type ChangedParamType = Omit<ParamType, 'format' | '_isParamType'>;

function combinationsMap(
  inputs: ParamType[],
  indexedArgsCount: number,
  change: (param: ParamType) => ChangedParamType,
): ChangedParamType[][] {
  const combinations = generateCombinations(inputs.length, indexedArgsCount);
  return combinations.map((combination) =>
    inputs.map((input, idx) =>
      combination.includes(idx) ? change(input) : input,
    ),
  );
}
