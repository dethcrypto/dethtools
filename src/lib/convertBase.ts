import BigNumber from 'bignumber.js';

import { Base } from './convertBaseProperties';

export function convertBase(
  value: string,
  from: Base,
  to: Base,
): string | undefined {
  return convertBaseWithBn(value, to, baseNameToBase[from], baseNameToBase[to]);
}

// @internal
const baseNameToBase = {
  binary: 2,
  octal: 8,
  decimal: 10,
  hexadecimal: 16,
};

// @internal
const baseNameToPrefix: Partial<Record<Base, string>> = {
  octal: '0',
  hexadecimal: '0x',
};

// @internal
function convertBaseWithBn(
  value: string,
  toBaseName: Base,
  from: number,
  to: number,
): string {
  return (
    String(baseNameToPrefix[toBaseName] || '') +
    new BigNumber(value, from).toString(to)
  );
}
