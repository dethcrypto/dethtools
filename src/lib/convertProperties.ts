import { PrecisionDict } from './convertUnits';

export const unitPrecision: PrecisionDict<UnitType> = {
  wei: 0, // 1e0 === 1
  gwei: 9,
  eth: 18,
};

export type UnitType = 'wei' | 'gwei' | 'eth';

const unitTypes: UnitType[] = ['wei', 'gwei', 'eth'];

// eslint-disable-next-line @typescript-eslint/no-redeclare
export const UnitType = { values: unitTypes };

export type TokenUnitType = 'unit' | 'base';
