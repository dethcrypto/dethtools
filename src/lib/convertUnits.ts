import { BigNumber } from 'bignumber.js';

import { unitPrecision, UnitType } from './convertProperties';

BigNumber.set({ EXPONENTIAL_AT: 1000, DECIMAL_PLACES: 1000 });

export type PrecisionDict<T extends string> = { [key in T]: number };

export function convertEthUnits(
  value: string,
  fromUnit: UnitType,
  toUnit: UnitType,
): string | undefined {
  return convertUnit(value, fromUnit, toUnit, unitPrecision);
}

// @internal
export function convertUnit<T extends string>(
  value: string,
  fromUnit: T,
  toUnit: T,
  precisionDict: PrecisionDict<T>,
): string | undefined {
  return new BigNumber(value)
    .times(unitPrecisionToValue(precisionDict[fromUnit]))
    .div(unitPrecisionToValue(precisionDict[toUnit]))
    .toString();
}

// @internal
export function unitPrecisionToValue(precision: number): string {
  if (precision < 0) return '0';
  return ['1'].concat('0'.repeat(precision)).join('');
}
