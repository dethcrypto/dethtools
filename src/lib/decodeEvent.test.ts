import { Interface } from '@ethersproject/abi';
import { expect } from 'earljs';

import {
  DecodedEventResultWithArgNames,
  decodeEvent,
  doesContainNamedKeys,
  EventProps,
  omitNonNamedKeys,
} from './decodeEvent';
import { parseAbi } from './parseAbi';

describe(omitNonNamedKeys.name, () => {
  it('filters non named keys e.g 0, 1 correctly', () => {
    expect(
      omitNonNamedKeys({
        '0': 'hello',
        1: 'world!',
        dethTools: 'is',
        awesome: '8)',
      }),
    ).toEqual({
      dethTools: 'is',
      awesome: '8)',
    });
  });
});

describe(doesContainNamedKeys.name, () => {
  it('checks if object contains named keys', () => {
    expect(
      doesContainNamedKeys({
        '0': 'hello',
        1: 'world!',
        dethTools: 'is',
        awesome: '8)',
      }),
    ).toEqual(true);
    expect(doesContainNamedKeys({ '0': 'hello', 1: 'world!' })).toEqual(false);
  });
});

describe(decodeEvent.name, () => {
  it('handles expected case', () => {
    const iface = parseAbi(`
      'constructor(string symbol, string name)',
      'function transferFrom(address from, address to, uint256 amount)',
      'function getUser(uint256 id) view returns (tuple(string name, address addr) user)',
      'event Transfer(address indexed src, address indexed dst, uint256 amount)',
      `) as Interface;
    const eventProps: EventProps = {
      data: '0x0000000000000000000000000000000000000000000000000de0b6b3a7640000',
      topics: [
        '0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef',
        '0x0000000000000000000000008ba1f109551bd432803012645ac136ddd64dba72',
        '0x000000000000000000000000ab7c8803962c0f2f5bbbe3fa8bf41cd82aa1923c',
      ],
    };
    const decodeEventResult = decodeEvent(iface, eventProps);

    expect(decodeEventResult).toBeDefined();

    const { fullSignature } =
      decodeEventResult! as DecodedEventResultWithArgNames;

    expect(fullSignature).toEqual(
      'event Transfer(address indexed src, address indexed dst, uint256 amount)',
    );
    expect(decodeEventResult).toMatchSnapshot();
  });

  it('handles case with data without 0x prefix', () => {
    const iface = parseAbi(`
      'event Transfer(address indexed src, address indexed dst, uint256 amount)',
      `) as Interface;
    const eventProps: EventProps = {
      data: '0000000000000000000000000000000000000000000000000de0b6b3a7640000',
      topics: [
        '0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef',
        '0x0000000000000000000000008ba1f109551bd432803012645ac136ddd64dba72',
        '0x000000000000000000000000ab7c8803962c0f2f5bbbe3fa8bf41cd82aa1923c',
      ],
    };
    const decodeEventResult = decodeEvent(iface, eventProps);

    expect(decodeEventResult).toMatchSnapshot();
  });

  it('handles case with too few topics', () => {
    const iface = parseAbi(`
      'constructor(string symbol, string name)',
      'function transferFrom(address from, address to, uint256 amount)',
      'function getUser(uint256 id) view returns (tuple(string name, address addr) user)',
      'event Transfer(address indexed src, address indexed dst, uint256 amount)',
      `) as Interface;
    const eventProps: EventProps = {
      data: '0x0000000000000000000000000000000000000000000000000de0b6b3a7640000',
      topics: [
        '0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef',
        '0x0000000000000000000000008ba1f109551bd432803012645ac136ddd64dba72',
      ],
    };

    expect(() => decodeEvent(iface, eventProps)).toThrow(
      expect.stringMatching(/data out-of-bounds/),
    );
  });

  it('handles non-named params case', () => {
    const iface = parseAbi(`
      'constructor(string symbol, string name)',
      'function transferFrom(address from, address to, uint256 amount)',
      'function getUser(uint256 id) view returns (tuple(string name, address addr) user)',
      'event Transfer(address indexed, address indexed, uint256)',
      `) as Interface;
    const eventProps: EventProps = {
      data: '0x0000000000000000000000000000000000000000000000000de0b6b3a7640000',
      topics: [
        '0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef',
        '0x0000000000000000000000008ba1f109551bd432803012645ac136ddd64dba72',
        '0x000000000000000000000000ab7c8803962c0f2f5bbbe3fa8bf41cd82aa1923c',
      ],
    };
    const decodeEventResult = decodeEvent(iface, eventProps);

    const { hasArgumentNames } = decodeEventResult!;

    expect(hasArgumentNames).toEqual(false);
    expect(decodeEventResult).toMatchSnapshot();
  });

  it('handles case where abi does not contain events', () => {
    const iface = parseAbi(`
      'constructor(string symbol, string name)',
      'function transferFrom(address from, address to, uint256 amount)',
      'function getUser(uint256 id) view returns (tuple(string name, address addr) user)',
      `) as Interface;
    const eventProps: EventProps = {
      data: '0x0000000000000000000000000000000000000000000000000de0b6b3a7640000',
      topics: [
        '0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef',
        '0x0000000000000000000000008ba1f109551bd432803012645ac136ddd64dba72',
        '0x000000000000000000000000ab7c8803962c0f2f5bbbe3fa8bf41cd82aa1923c',
      ],
    };

    try {
      decodeEvent(iface, eventProps);
    } catch (err: any) {
      expect(err.reason).toEqual('no matching event');
      expect(err).toMatchSnapshot();
    }
  });
});
