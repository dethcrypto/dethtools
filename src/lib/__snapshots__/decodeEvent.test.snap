// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`decodeEvent handles case where abi does not contain events 1`] = `[Error: no matching event (argument="topichash", value="0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef", code=INVALID_ARGUMENT, version=abi/5.6.4)]`;
